Add support for subgrids and fix some other grid properties

wafflespeanut opened the pull request:
This covers the `subgrid <line-name-lists>?` for `grid-template` (for gecko, and hence, stylo).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/17268)
<!-- Reviewable:end -->

Commit:
Add support for subgrid line name lists

Commit:
Switch to GridTemplateComponent for subgrid line names

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/values/generics/grid.rs, components/style/values/specified/mod.rs, components/style/properties/gecko.mako.rs, components/style/values/computed/mod.rs, components/style/properties/shorthand/position.mako.rs and 2 more
 * @emilio: components/style/values/generics/grid.rs, components/style/values/specified/mod.rs, components/style/properties/gecko.mako.rs, components/style/values/computed/mod.rs, components/style/properties/shorthand/position.mako.rs and 2 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify style code, but no tests are modified. Please consider adding a test!

Commit:
Cleanup some old code

nox commented on components/style/values/generics/grid.rs:
```diff
@@ -556,3 +557,93 @@ impl<T: ToCss> ToCss for TrackList<T> {
+#[cfg_attr(feature = "servo", derive(HeapSizeOf))]
+pub struct LineNameList {
+    /// The optional `<line-name-list>`
+    pub names: Vec<Vec<CustomIdent>>,
```
I wonder if that should be `Box<[Box<[CustomIdent]>>`, but that's unrelated to that patch.

nox commented on components/style/values/generics/grid.rs:
```diff
@@ -556,3 +557,93 @@ impl<T: ToCss> ToCss for TrackList<T> {
+/// The `<line-name-list>` for subgrids.
+///
+/// `[ <line-names> | repeat(<positive-integer> | auto-fill, <line-names>+) ]+`
+/// **NOTE; This is deprecated!**
```
How is this deprecated? Didn't it just got moved to a different greater version of the spec?

nox commented on components/style/values/generics/grid.rs:
```diff
@@ -556,3 +557,93 @@ impl<T: ToCss> ToCss for TrackList<T> {
     }
 }
+
+/// The `<line-name-list>` for subgrids.
```
Spec link would be nice.

nox commented on components/style/values/specified/grid.rs:
```diff
@@ -349,3 +349,52 @@ impl ToComputedValue for TrackList<TrackSizeOrRepeat> {
+    }
+}
+
+impl ToComputedValue for GridTemplateComponent<TrackSizeOrRepeat> {
```
This can be derived.

wafflespeanut commented:
I'll address the comments once I'm convinced about the test failures.

wafflespeanut commented:
@canaltinova I just talked to @bholley. I won't be very active for at least a month (except in IRC and responding to comments). Could you take ownership of this PR? Feel free to ping me for anything that doesn't make sense (or look weird). Also, this was my last try run - https://treeherder.mozilla.org/#/jobs?repo=try&revision=2385382ec93c8f6dc480b5b06c1cecd25c6e1d48&selectedJob=108086299

canaltinova commented:
@wafflespeanut Sure, I'll work on that next.

canova committed:
Fix the serialization of TrackRepeat

It was printing `repeat(n, <line-names>)` for number TrackRepeat.
But it should just print n times <line-names> instead.

MozReview-Commit-ID: FEmL8O2osVb

canova committed:
Fall back to an older version of the spec in TrackRepeat for the parity with gecko

canova committed:
Prevent parsing 'none' keyword in grid-template's columns part

canova committed:
Fix parsing of grid-template shorthand

It should accept 'grid-template-rows / grid-template-columns values' values not
'grid-template-rows grid-template-columns values'.

canova committed:
Fix 'grid' shorthand serialization

canova committed:
Fix parsing logic of grid-template property

Fist line names are parsed on the top of the function already and it shouldn't be parsed again.

canova committed:
Fix parsing of 'subgrid' and 'none' keywords in grid-template

canova committed:
Fix serialization of grid-template

If both columns and rows are 'none', shorthand should serialize it as 'none / none'
instead of just 'none'.

canaltinova commented:
Latest try job: https://treeherder.mozilla.org/#/jobs?repo=try&revision=76c0777c8719950bccc2647e233544da9cdf131b We still need to fix the remaining 53 failures. But this looks like a good start. I think we can land this and deal with the rest later.

canaltinova commented:
@bors-servo r=wafflespeanut,canaltinova

bors-servo commented:
:pushpin: Commit e640194 has been approved by `wafflespeanut,canaltinova`

<!-- @bors-servo r=wafflespeanut,canaltinova e640194a8fb9f870844eb3872c37aa0cf9a715de -->

bors-servo commented:
:hourglass: Testing commit e640194a8fb9f870844eb3872c37aa0cf9a715de with merge 2d81bce47f8d7554bc222101d467c55c36ce5d8c...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/8116)

canaltinova commented:
@bors-servo r=wafflespeanut,canaltinova

bors-servo commented:
:pushpin: Commit aaa9170 has been approved by `wafflespeanut,canaltinova`

<!-- @bors-servo r=wafflespeanut,canaltinova aaa9170032a00f5522fa8aa9c2a375f116f52568 -->

bors-servo commented:
:hourglass: Testing commit aaa9170032a00f5522fa8aa9c2a375f116f52568 with merge eba99eb035488d828d2f22e90becd0da563ebc45...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/8117)

canaltinova commented:
@bors-servo r=wafflespeanut,canaltinova
(Oops, I thought that test can accept 2 parameters somehow.)

bors-servo commented:
:pushpin: Commit 280cf12 has been approved by `wafflespeanut,canaltinova`

<!-- @bors-servo r=wafflespeanut,canaltinova 280cf1217464f8b3477e9aeef162f50579abe6fb -->

bors-servo commented:
:hourglass: Testing commit 280cf1217464f8b3477e9aeef162f50579abe6fb with merge 88b4ff49950e3e040d0a3c95a648a0d8d92c14f3...

bors-servo commented:
:broken_heart: Test failed - [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3352)

canaltinova commented:
@bors-servo r=wafflespeanut,canaltinova
(more test changes)

bors-servo commented:
:pushpin: Commit 3e49a8c has been approved by `wafflespeanut,canaltinova`

<!-- @bors-servo r=wafflespeanut,canaltinova 3e49a8c80ab639cefd1ae3947deb16bbfdf60ae4 -->

bors-servo commented:
:hourglass: Testing commit 3e49a8c80ab639cefd1ae3947deb16bbfdf60ae4 with merge 7a71035793f6d87c8f0f8e63ae045c1aaa40b617...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/7774), [arm32](http://build.servo.org/builders/arm32/builds/7994), [arm64](http://build.servo.org/builders/arm64/builds/8030), [linux-dev](http://build.servo.org/builders/linux-dev/builds/8119), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/4874), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/4907), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/8108), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/587), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/584), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/4814), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/4871), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/585), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/590), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/3353)
Approved by: wafflespeanut,canaltinova
Pushing 7a71035793f6d87c8f0f8e63ae045c1aaa40b617 to master...
